### PR TITLE
Unfuck iBGP loop detection et all

### PR DESCRIPTION
**Pull request by BarbarossaTM**
Add tracking of contributing ASNs to a LocRIB.

Signed-off-by: Maximilian Wilhelm <max@sdn.clinic>

### CODE REVIEW

File: routingtable/contributing_asn_list.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
+package routingtable
```
Comment (taktv6) [inline routingtable/contributing_asn_list.go]: Please add tests for this file.

File: protocols/bgp/server/server.go [hunk 2]
```diff
@@ -7,9 +7,10 @@ import (
 	"strings"
 	"sync"
 
+	"github.com/bio-routing/bio-rd/routingtable/locRIB"
+
```
Comment (q3k) [inline protocols/bgp/server/server.go]: Remove newline

File: routingtable/BUILD.bazel [hunk 3]
```diff
@@ -25,6 +26,7 @@ go_test(
     name = "go_default_test",
     srcs = [
         "client_manager_test.go",
+	"contributing_asn_list_test.go",
```
Comment (q3k) [inline routingtable/BUILD.bazel]: Indent with spaces.

File: routingtable/contributing_asn_list.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+type contributingASN struct {
+	asn   uint32
+	count uint32
+}
+
+// ContributingASNs contains a list of contributing ASN to a LocRIB to check ASPaths for possible routing loops.
+type ContributingASNs struct {
+	contributingASNs []*contributingASN
+	mu               sync.RWMutex
```
Comment (q3k) [inline routingtable/contributing_asn_list.go]: This mutex should be commented/named so it's obvious what does it guard.

File: routingtable/contributing_asn_list.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+				c.contributingASNs = asnList[:len(asnList)-1]
+			}
+			return
+		}
+	}
+}
+
+// IsContributingASN checks if  a given ASN is part of the contributing ASNs
+func (c *ContributingASNs) IsContributingASN(asn uint32) bool {
+	for _, cASN := range c.contributingASNs {
```
Comment (q3k) [inline routingtable/contributing_asn_list.go]: Guard with mutex.

File: routingtable/contributing_asn_list.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+	c.contributingASNs = append(c.contributingASNs, &contributingASN{
+		asn:   asn,
+		count: 1,
+	})
+}
+
+// Remove a ASN to the list of contributing ASNs or decrement the ref count of an existing one.
+func (c *ContributingASNs) Remove(asn uint32) {
+	c.mu.RLock()
```
Comment (q3k) [inline routingtable/contributing_asn_list.go]: This should be a normal Lock/Unlock.

File: routingtable/contributing_asn_list.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+	c := &ContributingASNs{
+		contributingASNs: []*contributingASN{},
+	}
+
+	return c
+}
+
+// Add a new ASN to the list of contributing ASNs or add the ref count of an existing one.
+func (c *ContributingASNs) Add(asn uint32) {
+	c.mu.RLock()
```
Comment (q3k) [inline routingtable/contributing_asn_list.go]: This should be a normal Lock/Unlock

File: routingtable/contributing_asn_list.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+	defer c.mu.RUnlock()
+
+	for _, cASN := range c.contributingASNs {
+		if cASN.asn == asn {
+			cASN.count++
```
Comment (q3k) [inline routingtable/contributing_asn_list.go]: Check for overflow? panic() is fine if we do.

File: routingtable/contributing_asn_list.go [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+// Remove a ASN to the list of contributing ASNs or decrement the ref count of an existing one.
+func (c *ContributingASNs) Remove(asn uint32) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	asnList := c.contributingASNs
+
+	for i, cASN := range asnList {
+		if cASN.asn == asn {
```
Comment (q3k) [inline routingtable/contributing_asn_list.go]: Use a guard-style expression instead:

    for i, cASN := range asnList {
        if cASN.asn != asn {
            continue;
        }

        cASNcount--
        // ...
    }

File: protocols/bgp/server/peer.go [hunk 10]
```diff
@@ -5,6 +5,8 @@ import (
 	"sync"
 	"time"
 
+	"github.com/bio-routing/bio-rd/routingtable/locRIB"
+
```
Comment (q3k) [inline protocols/bgp/server/peer.go]: Remove newline.

File: routingtable/adjRIBIn/adj_rib_in.go [hunk 11]
```diff
@@ -14,16 +15,18 @@ import (
 // AdjRIBIn represents an Adjacency RIB In as described in RFC4271
 type AdjRIBIn struct {
 	routingtable.ClientManager
-	rt           *routingtable.RoutingTable
-	mu           sync.RWMutex
-	exportFilter *filter.Filter
+	rt               *routingtable.RoutingTable
```
Comment (q3k) [inline routingtable/adjRIBIn/adj_rib_in.go]: Nit: the naming of these is not consistent: `RoutingTable` gets shortened to `rt`, but then there's a long `exportFilter` and `constibutingASNs`...

File: routingtable/adjRIBIn/adj_rib_in.go [hunk 12]
```diff
@@ -65,12 +68,32 @@ func (a *AdjRIBIn) AddPath(pfx net.Prefix, p *route.Path) error {
...
 	return nil
 }
 
+func (a *AdjRIBIn) ourASNsInPath(p *route.Path) bool {
+	// Don't accept path via iBGP which contain our ASN
+	ASPathAttr, _ := packet.ParseASPathStr(p.BGPPath.ASPath)
+
+	for _, pathSegment := range ASPathAttr.Value.(packet.ASPath) {
+		for _, asn := range pathSegment.ASNs {
+			if a.contributingASNs.IsContributingASN(asn) {
```
Comment (q3k) [inline routingtable/adjRIBIn/adj_rib_in.go]: Hm, now that I think about it:

Not using a map for the contributingASNs list causes us to do an O(n^2) operation here. We really should make sure slices are faster for small cases...

Comment (q3k) [reply routingtable/adjRIBIn/adj_rib_in.go]: (no action required)

**Comment by q3k** [comment]
Thank you!

**Pull request merged**